securesystemslib/hsm.py

```diff
@@ -0,0 +1,446 @@
+
+    <Returns>
+      List of all available public key handles.
+    """
+
+    public_key_objects = self.session.findObjects([(PyKCS11.CKA_CLASS,
+        PyKCS11.CKO_PUBLIC_KEY)])
+
+    return public_key_objects
+
+
+
+
+
+  def get_public_key_value(self, public_key_handle):
+    """
+    <Purpose>
+      Get the public key value corresponding to the 'public_key_handle'
+
+    <Arguments>
+      public_key_handle:
+        element of the list returned by get_public_key_objects().
+
+    <Returns>
+      'cryptography' public key object
+    """
+
+    public_key_value = self.session.getAttributeValue(public_key_handle,
+        [PyKCS11.CKA_VALUE])[0]
+    public_key_value = bytes(public_key_value)
+
+    # Public key value exported from the HSM is der encoded
+    public_key = serialization.load_der_public_key(public_key_value,
+        default_backend())
+    return public_key
+
+
+
+
+
+  def get_X509_objects(self):
+    """
+    <Purpose>
+      Get object handle of the X509 certificates stored on the HSM.
+
+    <Returns>
+      List of all the available certificate handles.
+    """
+
+    x509_objects = self.session.findObjects([(PyKCS11.CKA_CLASS,
+        PyKCS11.CKO_CERTIFICATE)])
+    return x509_objects
+
+
+
+
+
+  def get_X509_value(self, x509_handle):
+    """
+    <Purpose>
+      Get the certificate value corresponding to the 'x509_handle'.
+
+    <Arguments>
+      x509_handle:
+        element from the list returned by get_X509_objects().
+
+    <Returns>
+      'cryptography' public key object.
+    """
+
+    x509_value = self.session.getAttributeValue(x509_handle,
+        [PyKCS11.CKA_VALUE])[0]
+    x509_certificate = x509.load_der_x509_certificate(
+        bytes(x509_value), default_backend())
+
+    return x509_certificate
+
+
+
+
+
+  def generate_signature(self, data, private_key_handle):
+    """
+    <Purpose>
+      Calculate signature over 'data' using the private key corresponding
+      to the 'private_key_handle'
+
+      Supported Keys
+      1. RSA
+      2. ECDSA
+
+    <Arguments>
+      data:
+        bytes over which the signature is to be calculated
+        'data' should be encoded/serialized before it is passed here.
+
+      private_key_handle:
+        element from the list returned by the get_private_key_objects()
+
+    <Exceptions>
+      securesystemslib.exceptions.UnsupportedAlgorithmError, when the
+      key type of the 'private_key_handle' is not supported
+
+    <Returns>
+      HEX string of the generated signature
+    """
+
+    mechanism = None
+    key_type = self.session.getAttributeValue(private_key_handle,
+        [PyKCS11.CKA_KEY_TYPE])[0]
+
+    if PyKCS11.CKK[key_type] == 'CKK_RSA':
+      mechanism = PyKCS11.RSA_PSS_Mechanism(RSA_PSS_MECH,
+          RSA_PSS_HASH_SHA256, RSA_PSS_MGF_SHA256, RSA_PSS_SALT_LENGTH)
+
+    elif PyKCS11.CKK[key_type] == 'CKK_EC' or PyKCS11.CKK[key_type] == 'CKK_ECDSA':
+      mechanism = PyKCS11.Mechanism(PyKCS11.CKM_ECDSA)
```

Comment:
The mechanism to be used should be negotiated with the slot (provided by [PyKCS11Lib.getMechanismList](https://pkcs11wrap.sourceforge.io/api/api.html#PyKCS11.PyKCS11Lib.getMechanismList)) to ensure that a supported one gets used. Mine, for example, doesn't support any of the PSS mechanisms.

I'd rather provide a sorted list (being the first the preferred mechanism) and automatically select the first matching one.

BTW this is exactly the same code as in lines 384-389, right? I'd rather extract that code to a function, not only for code economy, but in order to ensure that the same negotiation algorithm gets used.

Reply:
To get started with the work, We kept it simple at first (worked on the issue according to the mechanisms available with the YubiKey). I can make the required changes to get it supported with various HSM available. 

How would we decide the list of mechanisms to be made available?